src/lib/releaseTestPlanGenerator.ts

```diff
@@ -0,0 +1,171 @@
+import parse from "csv-parse/lib/sync";
+import { exportTestPlan } from "./testPlanFormatter";
+import {
+   findAllFilesForPattern,
+   getFilename as getFilenameFromCsv,
+   readContentFromFile,
+} from "./fileUtilities";
+import {
+   TestCase,
+   TestPlan,
+   testCaseExists,
+   findIndexForTestCase,
+   sortTestCases,
+} from "./testPlan";
+
+const RECORD_DELIMITER = ["\n", "\r\n", "\r"];
+
+/**
+ * Iterate over multiple feature test plans to extract all test cases
+ * and export them into a consolidated release test plan (csv format)
+ * @param filename
+ * @param versionToTest
+ * @param testPlanDir
+ */
+export async function generateReleaseTestPlan(
+   filename: string,
+   versionToTest: string,
+   testPlanDir: string
+): Promise<string> {
+   const files = await findTestPlanFiles(testPlanDir);
+
+   const testPlans = await Promise.all(
+      files.map(filename => loadTestPlanFromFile(filename))
+   );
+
+   // Multiple test plans might require the same test case
+   // which requires merging to consolidate their required status
+   const mergedTestCases: TestCase[] = [];
+
+   testPlans
+      .filter(testPlan => !!testPlan.testCases)
+      .map(testPlan => {
+         testPlan.testCases.map(testCase =>
+            mergeTestCase(
+               testCase,
+               getFilenameFromCsv(testPlan.name),
+               mergedTestCases
+            )
+         );
+      });
+
+   const mergedTestPlan: TestPlan = {
+      testCases: mergedTestCases,
+      name: filename,
+      versionToTest: versionToTest,
+   };
+   return await exportTestPlan(mergedTestPlan);
+}
+
+/**
+ * Merge a Test Case when it has already been required by another Test Plan
+ */
+export function mergeTestCase(
+   testCase: TestCase,
+   requiredByTestPlanName: string,
+   mergedTestCases: TestCase[]
+): void {
+   // Test case does not exists, add it
+   if (!testCaseExists(testCase, mergedTestCases)) {
+      var newRequiredBy = testCase.isRequired ? requiredByTestPlanName : "";
+
+      mergedTestCases.push({
+         featureName: testCase.featureName,
+         isRequired: testCase.isRequired,
+         requiredBy: newRequiredBy,
+      });
+   } else {
+      // Test case exists, requires merging
+      const existingTestCaseIndex = findIndexForTestCase(
```

Comment:
Seems like it would be more straightforward to get the testCase instead of the index; then the object could just be mutated directly

Reply:
Done